File: java/unittests/set/Set.java
/*
 * NAME: Set
 * AUTHOR: Tanaka Chitete
 * PURPOSE: Implement generic Set
 * CREATION: 22/09/2020
 * LAST MODIFICATION: 02/03/2021
 */

import java.util.*;

public class Set<E> extends AbstractSet<E> {
    // PRIVATE CLASS FIELDS

    private HashMap<E, E> entries;

    // CONSTRUCTORS

    /*
     * CONSTRUCTOR
     * IMPORT(S): NONE
     * EXPORT(S): Address of new Set
     * PURPOSE: Create new Set
     * CREATION: 22/10/2020
     * LAST MODIFICATION: 02/03/2021
     */

    public Set() {
        this.entries = new HashMap<E, E>();
    }

    // SETTERS (MUTATORS)

    @Override
    public boolean add(E element) {
        this.entries.put(element, element);
        return true; // put() always associates K with V
    }

    // GETTERS (ACCESSORS)

    @Override
    public boolean contains(Object element) {
        return this.entries.containsKey(element);
    }

    @Override
    public boolean remove(Object element) {
        return this.entries.remove(element, element);
    }

    @Override
    public int size() {
        return this.entries.size();
    }

    // OPERATORS

    @Override
    public Iterator<E> iterator() {
        throw new UnsupportedOperationException();
    }

    @Override
    public String toString() {
        return this.entries.values().toString();
    }

    @Override
    public boolean isEmpty() {
        return this.entries.isEmpty();
    }
}

File: python/set.py
"""
NAME: Set
AUTHOR: Tanaka Chitete
PURPOSE: Implement Set
CREATION: 02/03/2021
LAST MODIFICATION: 02/03/2021
"""

class Set:
    """
    CONSTRUCTOR
    IMPORT(S): NONE
    EXPORT(S): Address of new Set
    PURPOSE: Make new Set
    CREATION: 02/03/2021
    LAST MODIFICATION: 02/03/2021
    """

    def __init__(self, *args):
        self.entries = {}
        for a in args:
            self.add(a)


    # SETTERS (MUTATORS)

    def add(self, element):
        self.entries[element] = element


    # GETTERS (ACCESSORS)

    def has(self, element):
        return element in self.entries


    def remove(self, element):
        del self.entries[element]


    def __len__(self):
        return len(self.entries)


    def isEmpty(self):
        return not self.entries


    # OPERATORS

    def __str__(self):
        return str(list(self.entries.keys()))
